web/packages/teleport/src/Discover/SelectResource/SelectResource.tsx

```diff
@@ -335,67 +346,135 @@ function sortResourcesByKind(
   return sorted;
 }
 
+const aBeforeB = -1;
+const aAfterB = 1;
+const aEqualsB = 0;
+
+/**
+ * Evaluates the predicate and prioritizes the element matching the predicate over the element that
+ * doesn't.
+ *
+ * @example
+ * comparePredicate({color: 'green'}, {color: 'red'}, (el) => el.color === 'green') // => -1 (a before b)
+ * comparePredicate({color: 'red'}, {color: 'green'}, (el) => el.color === 'green') // => 1  (a after  b)
+ * comparePredicate({color: 'blue'}, {color: 'pink'}, (el) => el.color === 'green') // => 0  (both are equal)
+ */
+function comparePredicate<ElementType>(
+  a: ElementType,
+  b: ElementType,
+  predicate: (resource: ElementType) => boolean
+): -1 | 0 | 1 {
+  const aMatches = predicate(a);
+  const bMatches = predicate(b);
+
+  if (aMatches && !bMatches) {
+    return aBeforeB;
+  }
+
+  if (bMatches && !aMatches) {
+    return aAfterB;
+  }
+
+  return aEqualsB;
+}
+
 export function sortResources(
   resources: ResourceSpec[],
-  preferences: UserPreferences
+  preferences: UserPreferences,
+  onboardDiscover: OnboardDiscover | undefined
 ) {
   const { preferredResources, hasPreferredResources } =
     getPrioritizedResources(preferences);
+  const platform = getPlatform();
 
   const sortedResources = [...resources];
   const accessible = sortedResources.filter(r => r.hasAccess);
   const restricted = sortedResources.filter(r => !r.hasAccess);
 
-  // Sort accessible resources by 1. os 2. preferred 3. guided and 4. alphabetically
+  const hasNoResources = onboardDiscover && !onboardDiscover.hasResource;
```

Comment:
I just realized that this is probably not going to work, because `onboardDiscover.hasResource` is not actively updated. We just fetch the value for this field once if `onboardDiscover` is not in localStorage. This works fine for the Discover modal but doesn't work well for my use case.

Reply:
Wait, no, it actually should work but I just don't understand when we clear localStorage. It looks like we do so during logout, but does that mean that if my session naturally expires and I open the Web UI after that, localStorage gets cleared as well?

Reply:
yes, it'll get cleared on logout. when session becomes [invalid](https://github.com/gravitational/teleport/blob/632ba90530e047b5fbd733482a6188c20c86e704/web/packages/teleport/src/components/Authenticated/Authenticated.tsx#L51), we also clear it (and re-direct user to login). the call to `session.logout` clears the storage.

and you're right that it doesn't get actively updated (unless you refresh the browser), but i don't think it matters too much here.